Migrate from @noble/secp256k1 to @noble/curves for enhanced security

landabaso opened the pull request:
### Overview
This PR replaces the dependency `@noble/secp256k1` with `@noble/curves` for enhanced security and maintainability, following the advice of @paulmillr (https://github.com/bitcoinerlab/secp256k1/issues/6#issuecomment-2525124149). The migration ensures compatibility with the existing API.

### Changes
- Replaced all references to `@noble/secp256k1` with the appropriate modules from `@noble/curves`.

### Impact
- ~~**No breaking changes**: The public API remains unchanged, ensuring a seamless upgrade for users.~~
- **Updated behavior**: The `signSchnorr` function no longer defaults to a zero-filled `auxRand` for compatibility reasons. It now requires an explicit value or uses a secure random value internally.
- **Better maintainability**: By following the latest recommendations from the library's author, the codebase stays in sync with upstream improvements.

### Version
- Bumped version from `1.1.1` to `1.2.0` to indicate the significant internal change while keeping the same external functionality.

### Testing
- All existing tests pass with the updated dependency.

landabaso committed:
prettier code format

landabaso committed:
refactor: replace @noble/secp256k1 with @noble/curves for improved security and maintainability

- Replaced `@noble/secp256k1` dependency with `@noble/curves` as per the recommendation from @paulmillr.
- Updated all internal references to utilize `@noble/curves` APIs.
- Preserved the same API for external users while improving internal functionality.
- Bumped version to `1.2.0` to reflect the significance of the dependency change and internal improvements.

paulmillr commented on index.js:
```diff
@@ -109,84 +110,85 @@ function isExtraData(e) {
-  const Q = necc.Point.BASE.multiplyAndAddUnsafe(P, t, BigInt(1));
-  if (!Q) throw new Error('Tweaked point at infinity');
+  // multiplyAndAddUnsafe(P, scalar, 1) = P + scalar*G
+  const Q = Point.BASE.multiplyAndAddUnsafe(P, t, 1n);
```
BigInt(1) if you want to not use literals (compat)

paulmillr commented on index.js:
```diff
@@ -109,84 +110,85 @@ function isExtraData(e) {
+function _pointMultiply(p, tweak, isCompressed) {
+  const P = fromHex(p);
+  const h = typeof tweak === "string" ? tweak : utils.bytesToHex(tweak);
   const t = BigInt(`0x${h}`);
```
Better to use `hexToNumber`

paulmillr commented on index.js:
```diff
@@ -197,10 +199,39 @@ function throwToNull(fn) {
+  return Fp.add(Fp.add(x3, Fp.mul(x, a)), b); // x3 + a * x + b
+}
+
+function fromXOnly(bytes) {
```
There is `schnorr.utils.lift_x(bytesToNumber(bytes))`

paulmillr commented on index.js:
```diff
@@ -197,10 +199,39 @@ function throwToNull(fn) {
+  const x = utils.bytesToNumberBE(bytes);
+  if (!isValidFieldElement(x)) throw new Error("Point is not on curve");
+  const y2 = weierstrassEquation(x); // y² = x³ + ax + b
+  let y = Fp.sqrt(y2);
```
Since this doesn't use built-in check, it should be comparing `y * y === y2`

paulmillr replied:
But just use `schnorr.utils.lift_x(bytesToNumber(bytes))`

paulmillr commented on index.js:
```diff
@@ -356,11 +387,14 @@ export function signRecoverable(h, d, e) {
 }
 
-export function signSchnorr(h, d, e = Buffer.alloc(32, 0x00)) {
+export function signSchnorr(h, d, e = new Uint8Array(32)) {
```
Why are you filling `e` with zeros by default?? This massively decreases signature security and is against bip340 spec which mandates RANDOM e

landabaso replied:
This behavior was implemented to match `@bitcoinjs-lib/tiny-secp256k1`, as the aim of `bitcoinerlab/secp256k1` is to be a drop-in replacement for it. Users previously raised concerns about [[differences between the libraries](https://github.com/bitcoinerlab/secp256k1/issues/3)](https://github.com/bitcoinerlab/secp256k1/issues/3).

If you recall, we discussed this topic in the past: https://github.com/paulmillr/noble-curves/issues/61

`bitcoinjs-lib/tiny-secp256k1` is aligned with `bitcoin-core/secp256k1`, which also exhibits this behavior. However, since `tiny-secp256k1` has recently switched to the Rust `secp256k1` crate, perhaps this has been updated. Could you confirm, @junderw?

landabaso replied:
After reviewing previous discussions, I believe the current implementation (using a default empty array, as in `function signSchnorr(h, d, e = new Uint8Array(32))`) is ok. In fact, the issue was that `bitcoinjs-lib` was not doing this, [[which has since been addressed](https://github.com/bitcoinjs/tiny-secp256k1/issues/118#issuecomment-1612648330)](https://github.com/bitcoinjs/tiny-secp256k1/issues/118#issuecomment-1612648330).

paulmillr replied:
Ok.

to clarify, libsecp doesn’t use it because they can’t access good randomness. We can. Sipa mentioned:

> Using random nonces is recommended and obviously better, but libsecp256k1 has no access to high-quality randomness; it's up to the caller to provide it.

paulmillr commented on index.js:
```diff
@@ -370,27 +404,31 @@ export function signSchnorr(h, d, e = Buffer.alloc(32, 0x00)) {
 
-  return necc.recoverPublicKey(h, signature, recoveryId, assumeCompression(compressed));
+  const s = secp256k1.Signature.fromCompact(signature);
+  s.recovery = recoveryId;
```
`secp256k1.Signature.fromCompact(signature).addRecoveryBit(recoveryId)`

landabaso committed:
Update README to reflect migration to noble-curves library

landabaso committed:
Refactor to incorporate @paulmillr's suggestions

Details:

Replaced custom hexToNumber function with utils.hexToNumber and extend usage of it to all parts of the code.
Avoid literals.
Simplified fromXOnly implementation by leveraging schnorr.utils.lift_x.
Updated Signature handling in recover to use addRecoveryBit.
Removed custom implementations replaced by noble-curves utilities.

landabaso commented:
Thank you for the suggestions, @paulmillr. I’ve applied all of them. I’ll keep this open a little longer to wait for feedback from @junderw regarding `signSchnorr`.

landabaso committed:
Remove unused constants and variables after recent refactor

junderw commented:
To be honest, I agree with @paulmillr here...

Anyone who is running unit tests on the determinism of the signing function (which makes no sense to me...) or if they can't afford the random number generation performance hit (which is small anyways), they can provide their own constant e if they'd like.

I agree that we should default to random values.

Perhaps mention it in the documentation as a low hanging fruit for people who want to squeeze every last drop of performance.

landabaso commented:
Thanks for the feedback, both of you. I'll remove the constant initialization to zeros and reference this conversation in the documentation. It was included solely to maintain compatibility, but this is a minor difference.

junderw commented:
> massively decreases

While I agree. This is a bit exaggerated.

Quoting BIP340, the language used to describe the potential attacks is much more muted than "massive"... but again, I agree.

![スクリーンショット 2024-12-12 15 29 23](https://github.com/user-attachments/assets/05cd6256-d1dd-453c-a3af-695105a7b723)

paulmillr commented:
@junderw one example why it’s important: there is a fault attack which leaks private keys, in one of popular js libs. It will be public soon. auxRandom would have totally prevented it.

landabaso committed:
Remove default zero initialization for e in signSchnorr, defaulting to random as per recommended practices.

landabaso committed:
Document signSchnorr behavior change for auxiliary randomness parameter in README

landabaso committed:
Clarify signSchnorr behavior deviation from bitcoinjs/tiny-secp256k1 in README

junderw commented:
I would support a PR that changes the wording of BIP340 to be a bit stronger in regards to auxRandom.

I think core would be more willing to change the wording once the details of the vulns are public.
